### PR TITLE
Shwetank tt/conv non tile sharding testing

### PR DESCRIPTION
**Pull request by shwetankTT**
Ticket
https://github.com/tenstorrent/tt-metal/issues/10110

Problem description
Supporting e non-tile sharding for height sharded tensors for convolution.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 1]
```diff
@@ -731,7 +742,7 @@ std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::T
     Conv2dConfig conv_config = conv_config_.value_or(Conv2dConfig());
     uint32_t output_height = ((input_height - kernel_size[0] - ((kernel_size[0] - 1 ) * (dilation[0] - 1)) + 2 * padding[0]) / stride[0]) + 1;
     uint32_t output_width = ((input_width - kernel_size[1] - ((kernel_size[0] - 1 ) * (dilation[0] - 1)) + 2 * padding[1]) / stride[1]) + 1;
-    auto [input_tensor_post_tm, parallel_config, tensor_manipulated] = shard_or_reshard_tensor_if_required(
+    auto [input_tensor_post_tm, parallel_config, tensor_manipulated, use_non_tile_height] = shard_or_reshard_tensor_if_required(
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:745]: Nice that the flag is no longer user provided. But do we still need both code paths?

**Review by mywoodstock — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 1 above)
Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:833]: @shwetankTT what is the resolution on this?

Comment (shwetankTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:835]: The flag is set if certain conditions are met. we need this flag to differentiate tile vs rows copy in kernel.

**Review by mtairum — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 2]
```diff
@@ -512,20 +527,18 @@ std::tuple<ttnn::Shape, ttnn::MemoryConfig, bool> get_conv_padded_input_shape_an
             1,
             input_tensor_height_snapped_to_tile,
             input_tensor_width_snapped_to_channels_alignment});  // TODO: resolve ttnn::types::Shape and
-                                                                 // tt::tt_metal::LegacyShape issue to clean up next line
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:515]: was removing this comment intentional?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 3]
```diff
@@ -476,8 +482,16 @@ std::tuple<ttnn::Shape, ttnn::MemoryConfig, bool> get_conv_padded_input_shape_an
     if (conv_config.reshard_if_not_optimal || needs_shard_or_reshard) {
         auto block_shard_orientation =
             conv_config.transpose_shards ? ShardOrientation::COL_MAJOR : ShardOrientation::ROW_MAJOR;
-        ParallelConfig optimal_parallel_config = determine_parallel_config(
-            shard_layout, batch_size, in_channels, height, width, out_channels, device, block_shard_orientation);
+        const ParallelConfig& optimal_parallel_config = determine_parallel_config(
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:485]: seems like a bad rebase/merge?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 2 above)
Comment (shwetankTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:515]: I might have removed it earlier will put it back.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 4]
```diff
@@ -431,6 +433,10 @@ std::tuple<ttnn::Shape, ttnn::MemoryConfig, bool> get_conv_padded_input_shape_an
             dilation,
             device);
     }
+    bool use_non_tile_height = shard_layout == TensorMemoryLayout::HEIGHT_SHARDED && out_channels <= 256 && conv_config.act_block_h_override == 0 &&
+        conv_config.dtype == DataType::BFLOAT16 && conv_config.output_layout == Layout::ROW_MAJOR;
+    use_non_tile_height = use_non_tile_height && conv_config.input_channels_alignment != 16; //shalow conv varient
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:438]: seems like this should just be part of the if statement above

File: models/demos/wormhole/llama31_8b/demo/demo_trace.py [hunk 5]
```diff
@@ -585,7 +585,7 @@ def run_llama_demo(user_input, batch_size, device, instruct_mode, is_ci_env, num
         "instruct_weights-3_batch",
     ],
 )
-@pytest.mark.parametrize("device_params", [{"trace_region_size": 7860224, "num_command_queues": 2}], indirect=True)
```
Comment (pavlejosipovic) [inline models/demos/wormhole/llama31_8b/demo/demo_trace.py:588]: is this related to this change?

File: tests/ttnn/unit_tests/operations/test_new_conv2d.py [hunk 6]
```diff
@@ -2296,3 +2304,116 @@ def test_conv_for_vanilla_unet(
...
+        (1, 32, 32, 240, 320, 3, 3, 1, 1, 1, 1, True, None),
+        (1, 64, 32, 240, 320, 3, 3, 1, 1, 1, 1, True, None),
+    ),
+)
+@pytest.mark.parametrize(
+    "weights_dtype",
+    [ttnn.bfloat8_b, ttnn.bfloat16],
+)
+@pytest.mark.parametrize(
+    "activations_dtype",
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/test_new_conv2d.py:2337]: do we now need separate batch of tests if feature is on by default?
does it get covered by existing tests?

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp [hunk 7]
```diff
@@ -142,8 +143,7 @@ std::vector<tt::tt_metal::LegacyShape> OptimizedConvNew::compute_output_shapes(c
     // Tiled output shape is padded shape. Padded to tile shape.
     auto shape_w = batch_size * conv_output_h * conv_output_w;
     auto shape_c = output_channels;
-    auto padded_shape_w =
-        parallelization_config.num_cores_nhw * parallelization_config.per_core_out_matrix_height_ntiles * TILE_HEIGHT;
+    auto padded_shape_w = this->use_non_tile_height ?  parallelization_config.num_cores_nhw * parallelization_config.per_core_out_matrix_height : parallelization_config.num_cores_nhw * parallelization_config.per_core_out_matrix_height_ntiles * TILE_HEIGHT;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:146]: pls format this line

File: models/demos/wormhole/llama31_8b/demo/demo_trace.py (hunk 5 above)
Comment (shwetankTT) [reply models/demos/wormhole/llama31_8b/demo/demo_trace.py:588]: Yeah. there was a failure in the pipeline when I ran this branch over the CI.

Comment (pavlejosipovic) [reply models/demos/wormhole/llama31_8b/demo/demo_trace.py:588]: ok, any idea how this change affects this?

File: tests/ttnn/unit_tests/operations/test_new_conv2d.py (hunk 6 above)
Comment (shwetankTT) [reply tests/ttnn/unit_tests/operations/test_new_conv2d.py:2337]: Currently it is not covered under the existing test cases. Test cases have been copied from resnet although resnet tests only covers tile layout.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 3 above)
Comment (shwetankTT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:485]: most probably auto formated during commit of this file

Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:485]: `const ParallelConfig& ` we don't need at const & part here

File: models/demos/wormhole/llama31_8b/demo/demo_trace.py (hunk 5 above)
Comment (shwetankTT) [reply models/demos/wormhole/llama31_8b/demo/demo_trace.py:588]: Yeah i ran the pipeline again after this change and it passed
